### PR TITLE
fix(async): safely overlap rollout weight sync

### PR DESCRIPTION
**Pull request by yicwang**
### Description
`actor.sync_weight_no_wait` previously affected only `AsyncEmbodiedRunner`.
`AsyncPPOEmbodiedRunner` is a sibling of that runner and always used the
blocking `EmbodiedRunner.update_rollout_weights()`, so the existing switch had
no effect for async PPO.

This change:

- adds an `AsyncWeightSyncMixin` shared by both async embodied runners;
- makes a background-sync handle complete only after rollout-side
  `WeightSyncer.apply(...)` has finished, rather than after the rollout worker
  merely acknowledges the request;
- keeps the initial actor-to-rollout sync blocking, drains an in-flight sync
  before a later blocking sync and normal teardown, and coalesces requests
  while one runner-level sync is in flight;
- permits no-wait synchronization only with the patch syncer. Bucket apply can
  yield between buckets, so allowing inference concurrently could expose a
  partially updated model;
- documents the behavior in English and Chinese and enables the no-wait path
  in the existing async-PPO ManiSkill e2e config.

The default remains `actor.sync_weight_no_wait=false`.

### Motivation and Context
Async PPO could not overlap actor-to-rollout weight transfer even though it
already exposed the same configuration as the other async embodied runner.
While sharing the implementation, two correctness gaps in the old async path
also needed to be closed: blocking transitions could race an outstanding
background apply, and normal teardown could stop workers before that apply had
finished.

Tracking rollout-side application is important for data correctness because
the applied model version is recorded on generated trajectories and used by
async PPO's staleness filtering.

### How has this been tested?
- `pytest -q tests/unit_tests/test_async_weight_sync.py`: **19 passed**. This
  covers both runner MROs, blocking/non-blocking behavior, runner and rollout
  coalescing, rollout-side apply completion, blocking and teardown drains,
  blocking startup, and the patch-only guard.
- selected patch/bucket round-trip cases from
  `tests/unit_tests/test_weight_syncer.py`: **5 passed**.
- `pre-commit run --files <changed files>`: Ruff lint, Ruff format, and
  committer signoff all passed.

GPU smoke test: one `rlinf-0` node with 4x NVIDIA H20 GPUs available; collocated
actor/env/rollout placement used GPU ranks 0-1. Pi0.5 + ManiSkill async PPO,
patch syncer, 4 environments, seed 0, 6 optimizer steps. The same config was
run with only `actor.sync_weight_no_wait` changed:

| | `false` | `true` |
| --- | ---: | ---: |
| Exit status | 0 | 0 |
| Actor versions | 1 through 6 | 1 through 6 |
| Maximum trajectory staleness observed | 2 | 2 |
| Discarded unused trajectories | 0 | 0 |
| Finite TensorBoard scalars | 274 / 274 | 285 / 285 |
| Mean runner-visible `update_rollout_weights` time | 2.621 s | 0.648 ms |

The no-wait run consumed trajectories tagged with rollout-applied versions
2-5 after initially consuming version 0, showing that applied versions—not
only actor-side counters—propagate into training data. This short run checks
the synchronization and staleness invariants plus numerical health; it is not
a convergence comparison.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/runners/async_embodied_runner.py [hunk 1]
```diff
@@ -308,6 +273,9 @@ def run(self):
             if profiled_step is not None:
                 self._close_profiling_window(profiled_step)
 
+        # Let any in-flight non-blocking sync land before the workers go away.
+        self.drain_pending_rollout_weight_sync()
```
Comment (guozhen1997) [inline rlinf/runners/async_embodied_runner.py:279]: Drain waits on `request_actor_sync_model` + `sync_model_to_rollout` handles. `request_actor_sync_model` returns after setting a flag, and actor `sync()` returns after send — neither waits for `WeightSyncer.apply()`. So this does not actually wait until rollout weights are applied. Please complete the rollout handle only after apply finishes.

File: rlinf/runners/async_embodied_runner.py
Comment (guozhen1997) [inline rlinf/runners/async_embodied_runner.py]: Commit says the initial sync stays blocking. PPO does that; this still passes `no_wait=self.sync_weight_no_wait`. Please call `self.update_rollout_weights()` here (blocking), and only use `no_wait` inside the loop.

File: rlinf/runners/async_ppo_embodied_runner.py [hunk 2]
```diff
@@ -159,7 +156,7 @@ def run(self) -> None:
                 self.global_step += 1
                 self.actor.set_global_step(self.global_step).wait()
                 with self.timer("update_rollout_weights"):
-                    self.update_rollout_weights()
+                    self.update_rollout_weights(no_wait=self.sync_weight_no_wait)
```
Comment (guozhen1997) [inline rlinf/runners/async_ppo_embodied_runner.py:159]: This is what first honors the flag for async PPO. The PR text says no-wait should be patch-only (bucket apply can yield a half-updated model). Please guard that here or in the rollout worker before recommending this path. Trajectory versions are updated only after apply.

File: rlinf/runners/async_embodied_runner.py (hunk 1 above)
Comment (yicwang) [reply rlinf/runners/async_embodied_runner.py:279]: request_actor_sync_model now keeps its coroutine alive until the rollout worker has applied the weights, so the handle is a real completion signal.

File: rlinf/runners/async_embodied_runner.py
Comment (yicwang) [reply rlinf/runners/async_embodied_runner.py]: Fixed.

File: rlinf/runners/async_ppo_embodied_runner.py (hunk 2 above)
Comment (yicwang) [reply rlinf/runners/async_ppo_embodied_runner.py:159]: Added the guard, but in AsyncWeightSyncMixin.init_weight_sync_state rather than at the call site, so it fails at startup instead of at the first sync. Happy to move it into the rollout worker if you'd prefer it closer to the apply path.

File: rlinf/runners/async_weight_sync.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# Copyright 2025 The RLinf Authors.
```
Comment (anHappyDog) [inline rlinf/runners/async_weight_sync.py]: shoud change to 2026.

File: rlinf/runners/async_weight_sync.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        # the same config key.
+        self.sync_weight_no_wait = self.cfg.actor.get("sync_weight_no_wait", False)
+        if self.sync_weight_no_wait:
+            weight_syncer_type = self.cfg.get("weight_syncer", {}).get("type", None)
+            if weight_syncer_type != "patch":
+                raise ValueError(
+                    "actor.sync_weight_no_wait=true requires "
+                    "weight_syncer.type=patch so a rollout cannot observe a "
+                    "partially applied bucket sync."
+                )
```
Comment (anHappyDog) [inline rlinf/runners/async_weight_sync.py]: it might be better to be put into `validate_embodied_cfg`.

File: rlinf/runners/async_ppo_embodied_runner.py [hunk 5]
```diff
@@ -159,7 +156,7 @@ def run(self) -> None:
...
                 self.rollout.set_global_step(self.global_step).wait()
```
Comment (anHappyDog) [inline rlinf/runners/async_ppo_embodied_runner.py]: if `sync_weight_no_wait` is True, rollout might set newer global step before rollout-side weight application finishes. Maybe it's better to delete this line here.

File: rlinf/runners/async_weight_sync.py (hunk 3 above)
Comment (anHappyDog) [inline rlinf/runners/async_weight_sync.py]: the file could be renamed to `async_weight_sync_mixin.py`？

Comment (yicwang) [reply rlinf/runners/async_weight_sync.py]: Fixed

File: rlinf/runners/async_ppo_embodied_runner.py (hunk 5 above)
Comment (yicwang) [reply rlinf/runners/async_ppo_embodied_runner.py]: Fixed.

File: rlinf/runners/async_weight_sync.py (hunk 4 above)
Comment (yicwang) [reply rlinf/runners/async_weight_sync.py]: Added the call to the end of validate_embodied_cfg().

**Review by anHappyDog — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]

**Comment by guozhen1997** [comment]
Hi @yicwang, could you please rebase this PR onto the latest main branch?

**Comment by yicwang** [comment]
> Hi @yicwang, could you please rebase this PR onto the latest main branch?

Done

**Comment by guozhen1997** [comment]
Hi @yicwang, could you please fix the filter coverage test?
```
The following changed files are not covered by any CI filter:
rlinf/runners/async_weight_sync_mixin.py
```

**Review by andylin-hao — APPROVED** [review]
